include/vigra/threadpool.hxx

```diff
@@ -0,0 +1,328 @@
+}
+
+
+
+template<class ITER, class F>
+void parallel_foreach_impl(
+    ThreadPool & pool,
+    const uint64_t nItems,                   
+    ITER iter, 
+    ITER end, 
+    F && f,
+    std::random_access_iterator_tag
+){
+    // typedef typename std::iterator_traits<ITER>::reference ReferenceType;
+    uint64_t workload = std::distance(iter, end);
+    const float workPerThread = float(workload)/pool.nThreads();
+    const uint64_t chunkedWorkPerThread = std::max(uint64_t(std::floor(workPerThread/3.0f+0.5f)), uint64_t(1));
+
+    for( ;iter<end; iter+=chunkedWorkPerThread){
+
+        // ITER localEnd = iter+chunkedWorkPerThread < end ? iter + chunkedWorkPerThread : end;
+        const size_t lc = std::min(workload, chunkedWorkPerThread);
+        workload-=lc;
+        pool.enqueue(
+            [&f, iter, lc]
+            (int id)
+            {
+                for(size_t i=0; i<lc; ++i)
+                    f(id, iter[i]);
+            }
+        );
+    }
+    pool.waitFinished();
+}
+
+
+
+template<class ITER, class F>
+void parallel_foreach_impl(
+    ThreadPool & pool,
+    const uint64_t nItems,                   
+    ITER iter, 
+    ITER end, 
+    F && f,
+    std::forward_iterator_tag
+){
+
+    // typedef typename std::iterator_traits<ITER>::reference ReferenceType;
+    uint64_t workload = nItems;
+    const float workPerThread = float(workload)/pool.nThreads();
+    const uint64_t chunkedWorkPerThread = std::max(uint64_t(std::round(workPerThread/3.0f)),1ul);
+
+    for(;;){
+
+        const size_t lc = std::min(chunkedWorkPerThread, workload);
+        workload -= lc;
+        pool.enqueue(
+            [&f, iter, lc]
+            (int id)
+            {
+                auto iterCopy = iter;
+                for(size_t i=0; i<lc; ++i){
+                    f(id, *iterCopy);
+                    ++iterCopy;
+                }
+            }
+        );
+        if(workload==0)
+            break;
+        std::advance(iter, lc);
+    }
+    pool.waitFinished();
+}
+
+
+
+template<class ITER, class F>
+void parallel_foreach_single_thread(
+    const uint64_t nItems,
+    ITER begin, 
+    ITER end, 
+    F && f
+){
+    for(size_t i=0; i<nItems; ++i){
+        f(0, *begin);
+        ++begin;
+    }
+}
+
+
+
+template<class ITER, class F>
+void parallel_foreach(
+    ThreadPool & pool,
+    const uint64_t nItems,
+    ITER begin, 
+    ITER end, 
+    F && f
+){
+
+    if(pool.nThreads()!=1){
+        parallel_foreach_impl(pool,nItems, begin, end, f,
+            typename std::iterator_traits<ITER>::iterator_category());
+    }
+    else{
+        parallel_foreach_single_thread(nItems, begin, end, f);
+    }
+}
+
+
+
+template<class ITER, class F>
+void parallel_foreach(
+    const int64_t nThreads,                  
+    const uint64_t nItems,
+    ITER begin, 
+    ITER end, 
+    F && f
+){
+    ThreadPool pool(nThreads==-1 ?  std::thread::hardware_concurrency() : nThreads);
+
+    if(pool.nThreads()!=1){
```

Comment:
Call this other overload instead of repeating the code. Both overloads should probably be inline.